### PR TITLE
add internal prometheus metrics

### PR DESCRIPTION
**Pull request by tgross**
For https://github.com/joyent/containerpilot/issues/374

This PR gives us a lot of the items described in issue #374, because most of the things that ContainerPilot does pass through the event bus `Publish` method. The list of events this will catch are:

- *ExitSuccess*: any `exec` (job or health check) completes with 0 exit code
- *ExitFailed*: any `exec` (job or health check) completes with 0 exit code
- *Stopping*: a job in this container is about to stop
- *Stopped*: a job in this container has stopped
- *StatusHealthy*: a job (or watched job) is healthy
- *StatusUnhealthy*: a job (or watched job) is unhealthy
- *StatusChanged*: a job (or watched job) status has changed
- *EnterMaintenance*: entered maintenance mode
- *ExitMaintenance*: exited maintenance mode
- *Error*: an `exec` couldn't start or was forced to cancel with an error b/c it timed out (will always have *ExitFailed* too, but *ExitFailed* might not include *Error*).
- *Metric*: a user-defined metric was recorded
- *Startup*: global ContainerPilot startup
- *Shutdown*: global ContainerPilot shutdown

(The internal *TimerExpired* and *Quit* events you can find in [events/events.go](https://github.com/joyent/containerpilot/blob/master/events/events.go) are never broadcast across the event bus so won't show up here.)

This Prometheus collector here is a vec counter, or in other words a counter that's annotated with labels. The output looks like this:

```
curl -s localhost:9090/metrics | grep containerpilot
# HELP containerpilot_events count of ContainerPilot events, partitioned by type and source
# TYPE containerpilot_events counter
containerpilot_events{code="ExitSuccess",source="check.app1"} 3
containerpilot_events{code="Startup",source="global"} 1
containerpilot_events{code="StatusHealthy",source="app1"} 3
```

This PR needs tests too, but I wanted to use this for discussion.
cc @cheapRoc @misterbisson @geek

### CODE REVIEW

**Comment by tgross** [comment]
I've added 849e397 to this PR, which adds a recording of HTTP status code and path for each control socket endpoint request. I'm not 100% sure this metric is at all useful though? Thoughts?

**Comment by tgross** [comment]
Ok, so the results of these collectors will now look like so:

```
curl -s localhost:9090/metrics | grep containerpilot
# HELP containerpilot_control_http_requests count of requests to control socket, partitioned by path and HTTP code
# TYPE containerpilot_control_http_requests counter
containerpilot_control_http_requests{code="200",path="/v3/ping"} 1
# HELP containerpilot_events count of ContainerPilot events, partitioned by type and source
# TYPE containerpilot_events counter
containerpilot_events{code="ExitSuccess",source="check.app1"} 1
containerpilot_events{code="Startup",source="global"} 1
containerpilot_events{code="StatusHealthy",source="app1"} 1
# HELP containerpilot_watch_instances gauge of instances found for each ContainerPilot watch, partitioned by service
# TYPE containerpilot_watch_instances gauge
containerpilot_watch_instances{service="otherapp"} 0
```

File: events/bus.go [hunk 1]
```diff
@@ -100,6 +111,7 @@ func (bus *EventBus) Publish(event Event) {
 	bus.lock.Lock()
 	defer bus.lock.Unlock()
 	log.Debugf("event: %v", event)
+	collector.WithLabelValues(event.Code.String(), event.Source).Inc()
```
Comment (cheapRoc) [inline events/bus.go]: So this is the only collector required for tracking events flowing through the bus? I can dig it.

File: control/control.go [hunk 2]
```diff
@@ -8,13 +8,25 @@ import (
...
+	"github.com/joyent/containerpilot/events"
 )
 
 // SocketType is the default listener type
 var SocketType = "unix"
 
+var collector *prometheus.CounterVec
+
+func init() {
+	collector = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (cheapRoc) [inline control/control.go]: I think I see what you mean now. In theory you could define all of your collectors or even just your prometheus option structs in a separate file and import them into the locations where you setup the collectors. I'm guessing thats the idea you mentioned in Slack but then went with placing each collector across the app.

On the opposite side, it's three definitions across this PR, one of which covers 80% of the events. LGTM.

**Comment by tgross** [comment]
Added an integration/smoke test to make sure we're picking up the collectors:

```
2017/07/06 16:43:52 ==== Running Integration Tests ====
2017/07/06 16:43:52 TEST: test_telemetry
Creating testtelemetry_consul_1 ...
Creating testtelemetry_consul_1 ... done
testtelemetry_consul_1 is up-to-date
Creating testtelemetry_app_1 ...
Creating testtelemetry_app_1 ... done
containerpilot_app_some_counter 42
# HELP containerpilot_events count of ContainerPilot events, partitioned by type and source
# TYPE containerpilot_events counter
containerpilot_events{code="ExitSuccess",source="check.app"} 1
containerpilot_events{code="ExitSuccess",source="sensor"} 1
containerpilot_events{code="Metric",source="containerpilot_app_some_counter|42"} 1
containerpilot_events{code="Startup",source="global"} 1
containerpilot_events{code="StatusHealthy",source="app"} 1
containerpilot_events{code="Stopped",source="sensor"} 1
containerpilot_events{code="Stopping",source="sensor"} 1
# HELP containerpilot_control_http_requests count of requests to control socket, partitioned by path and HTTP code
# TYPE containerpilot_control_http_requests counter
containerpilot_control_http_requests{code="200",path="/v3/metric"} 1
# HELP containerpilot_watch_instances gauge of instances found for each ContainerPilot watch, partitioned by service
# TYPE containerpilot_watch_instances gauge
containerpilot_watch_instances{service="otherapp"} 0
{"Services":[{"Name":"app","Address":"192.168.16.3","Port":8000,"Status":"unknown"},{"Name":"containerpilot","Address":"192.168.16.3","Port":9090,"Status":"healthy"}],"Watches":["otherapp"]}
        "ServiceID": "containerpilot-4545a3854855",
        "ServiceName": "containerpilot",
Stopping testtelemetry_app_1 ... done
[]
Going to remove testtelemetry_app_1, testtelemetry_test_run_1
Removing testtelemetry_app_1 ... done
Removing testtelemetry_test_run_1 ... done
2017/07/06 16:44:07 PASS: test_telemetry
```

**Pull request merged**